File: 02 Sorting/Binary Search 2/ex01e3.cpp
#include<iostream>
#include<vector>
using namespace std;
int N,M;

int binarySearch(vector<int> &numbers, int &target)
{
    int low = 0, high = numbers.size() - 1;
    int ans = -1;

    while(low <= high)
    {
        int mid = (low + high)/2;
        
        // Search right half, and store a number
        if(numbers[mid] <= target)
        {
            ans = numbers[mid];
            low = mid + 1;
        }
        // Search left half
        else high = mid-1;
    }
    return ans;
}

int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);
    
    // Input vectors
    cin >> N >> M;
    vector<int> numbers(N);
    for(int i = 0; i < N; i++) { cin >> numbers[i]; }

    while(M--)
    {
        // Input number
        int num;
        cin >> num;
        cout << binarySearch(numbers, num) << "\n";
    }
    return 0;
}

File: 02 Sorting/Binary Search 2/ex01e3.py
import sys
N,M = map(int, sys.stdin.readline().split())
def binarySearch(numbers, target):
    low = 0
    high = len(numbers) - 1
    ans = -1
    while(low <= high):
        mid = (low + high)//2
        if(numbers[mid] <= target):
            ans = numbers[mid]
            low = mid + 1
        else:
            high = mid - 1
    return ans
numbers = list(map(int, sys.stdin.readline().split()))
answers = [str(binarySearch(numbers, int(num))) 
            for num in sys.stdin.readline().split()]
sys.stdout.write("\n".join(answers))